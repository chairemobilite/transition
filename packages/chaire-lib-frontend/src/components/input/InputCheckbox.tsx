/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { JSX } from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import { _chunkify } from 'chaire-lib-common/lib/utils/LodashExtensions';

type choiceType = {
    value: string;
    iconPath?: string;
    label?: string;
    [key: string]: unknown;
};

type InputCheckboxCommonProps = WithTranslation & {
    id: string;
    localePrefix?: string;
    disabled?: boolean;
};

export type InputCheckboxProps = InputCheckboxCommonProps & {
    value?: string[];
    defaultValue?: string[];
    choices: choiceType[];
    columns?: number;
    onValueChange: (e: any) => void;
    allowSelectAll?: boolean;
};

type defaultInputType = {
    name: string;
    disabled?: boolean;
};

class InputCheckboxInner extends React.Component<InputCheckboxProps> {
    static defaultProps: Partial<InputCheckboxProps> = {
        defaultValue: [],
        choices: [],
        columns: 1,
        localePrefix: 'main',
        disabled: false,
        allowSelectAll: false
    };

    constructor(props: InputCheckboxProps) {
        super(props);
        this.onCheckboxInputChange = this.onCheckboxInputChange.bind(this);
        this.onContainerClick = this.onContainerClick.bind(this);
        this.selectAll = this.selectAll.bind(this);
    }

    onContainerClick(value, e) {
        e.stopPropagation();
        const clickEvent = new MouseEvent('click', {
            view: window,
            bubbles: true,
            cancelable: false
        });
        // this.onCheckboxInputChange({target: {value: value}})
    }

    onCheckboxInputChange(e) {
        e.stopPropagation();
        const values = this.props.value;

        const valueSet = new Set(values);
        let choices: string[] = [];
        const isChecked = e.target.checked;

        const checkboxValue = e.target.value;

        // use alt key as a check/uncheck all trigger
        if (serviceLocator && serviceLocator.keyboardManager && serviceLocator.keyboardManager.keyIsPressed('alt')) {
            if (valueSet.has(checkboxValue) && !isChecked) {
                // uncheck all
                choices = [];
            } else if (!valueSet.has(checkboxValue) && isChecked) {
                // check all
                const allChoices = this.props.choices.map((choice) => {
                    return choice.value;
                });
                choices = allChoices;
            }
        } else if (valueSet.has(checkboxValue) && !isChecked) {
            valueSet.delete(checkboxValue);
            choices = Array.from(valueSet);
        } else if (!valueSet.has(checkboxValue) && isChecked) {
            valueSet.add(checkboxValue);
            choices = Array.from(valueSet);
        }
        this.props.onValueChange({ target: { value: choices } });
    }

    onCheckboxClick(e) {
        e.stopPropagation();
    }

    onLabelClick(e) {
        e.stopPropagation();
    }

    selectAll(e: React.MouseEvent, checkAll: boolean): void {
        e.stopPropagation();
        const choices: string[] = checkAll ? this.props.choices.map((choice) => choice.value) : [];
        this.props.onValueChange({ target: { value: choices } });
    }

    render(): React.ReactNode {
        const values = this.props.value ? this.props.value : this.props.defaultValue ? this.props.defaultValue : [];
        const valueSet = new Set(values);
        const choices = this.props.choices;
        const localePrefix = this.props.localePrefix || 'main';
        const choiceInputs = choices.map((choice) => {
            const strValue =
                choice.value !== null && choice.value !== undefined ? choice.value.toString() : choice.value;
            const id = `${this.props.id}_${strValue}`;
            const strLabel = choice.label || this.props.t(`${localePrefix}:${choice.value}`) || choice.value;
            const iconPath = choice.iconPath;
            return (
                <div
                    key={id}
                    onClick={(e) => {
                        this.onContainerClick(choice.value, e);
                    }}
                    className={`tr__form-input-checkbox-container${valueSet.has(choice.value) ? ' checked' : ''}`}
                >
                    <div className="label-input-container">
                        <input
                            type="checkbox"
                            id={id}
                            name={this.props.id}
                            className={'_input-checkbox'}
                            value={strValue}
                            disabled={this.props.disabled === true ? true : false}
                            checked={valueSet.has(choice.value)}
                            onChange={this.onCheckboxInputChange}
                            onClick={this.onCheckboxClick}
                        />
                        <label htmlFor={id} onClick={this.onLabelClick}>
                            {iconPath && (
                                <span>
                                    <img src={iconPath} className="_icon" alt={strLabel} />
                                </span>
                            )}
                            <span>{strLabel}</span>
                        </label>
                    </div>
                </div>
            );
        }, this);

        // separate by columns if needed:
        let columnedChoiceInputs: JSX.Element[] = [];
        if (this.props.columns && this.props.columns > 0) {
            const widgetsByColumn =
                this.props.columns === 1 ? [choiceInputs] : _chunkify(choiceInputs, this.props.columns, true);
            for (let i = 0, count = widgetsByColumn.length; i < count; i++) {
                const columnWidgets = widgetsByColumn[i];
                columnedChoiceInputs.push(
                    <div
                        className={`tr__form-input-checkbox-group-column${this.props.columns === 1 ? ' no-wrap' : ''}`}
                        key={i}
                    >
                        {columnWidgets as JSX.Element[]}
                    </div>
                );
            }
        } else {
            columnedChoiceInputs = choiceInputs;
        }

        let selectAllWidget: JSX.Element | undefined = undefined;
        if (this.props.allowSelectAll && !this.props.disabled) {
            const allChecked = choices.find((choice) => !valueSet.has(choice.value)) === undefined;
            const id = `${this.props.id}_selectAll`;
            selectAllWidget = (
                <div className="tr__form-input-checkbox-group-column" key={id}>
                    <div className="label-input-container">
                        <input
                            type="button"
                            id={id}
                            name={this.props.id}
                            className={'_input-checkbox'}
                            value={
                                this.props.t(
                                    allChecked ? `${localePrefix}:UnselectAll` : `${localePrefix}:SelectAll`
                                ) as string
                            }
                            onClick={(e) => this.selectAll(e, !allChecked)}
                        />
                    </div>
                </div>
            );
        }

        return selectAllWidget ? (
            <div className={'apptr__form-input-container'}>
                {selectAllWidget}
                <div className={'tr__form-input-checkbox-group-container'}>{columnedChoiceInputs}</div>
            </div>
        ) : (
            <div className={'tr__form-input-checkbox-group-container'}>{columnedChoiceInputs}</div>
        );
    }
}

export const InputCheckbox = withTranslation(['main', 'transit'])(InputCheckboxInner);

export interface InputCheckboxBoolProps extends InputCheckboxCommonProps {
    isChecked?: boolean;
    defaultChecked?: boolean;
    label?: string;
    onValueChange: (e: any) => void;
}

class InputCheckboxBooleanInner extends React.Component<InputCheckboxBoolProps> {
    static defaultProps: Partial<InputCheckboxBoolProps> = {
        defaultChecked: false,
        label: 'Yes',
        onValueChange: (_e) => {
            /* nothing to do */
        },
        localePrefix: 'main',
        disabled: false
    };

    private _onValueChange: (e: any) => void;

    constructor(props: InputCheckboxBoolProps) {
        super(props);
        this.innerValueChange = this.innerValueChange.bind(this);
        this._onValueChange = props.onValueChange;
    }

    innerValueChange(e: any): void {
        if (e.target.value.includes('true')) {
            this._onValueChange({ target: { value: true } });
        } else {
            this._onValueChange({ target: { value: false } });
        }
    }

    render(): React.ReactNode {
        return (
            <InputCheckbox
                choices={[{ value: 'true', label: this.props.label }]}
                localePrefix={this.props.localePrefix}
                disabled={this.props.disabled}
                id={this.props.id}
                value={
                    this.props.isChecked
                        ? ['true']
                        : this.props.isChecked !== undefined
                            ? []
                            : this.props.defaultChecked
                                ? ['true']
                                : []
                }
                onValueChange={(e) => this.innerValueChange(e)}
            />
        );
    }
}

export const InputCheckboxBoolean = withTranslation('main')(InputCheckboxBooleanInner);

export default InputCheckbox;
