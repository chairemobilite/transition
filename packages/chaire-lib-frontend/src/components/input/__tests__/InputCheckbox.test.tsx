/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import { InputCheckbox, InputCheckboxBoolean } from '../InputCheckbox';
import { TestUtils } from 'chaire-lib-common/lib/test';

const mockOnChange = jest.fn();
const testId = 'SelectWidgetId';
const testLabel = 'Select label';

jest.mock('react-i18next', () => ({
    // this mock makes sure any components using the translate HoC receive the t function as a prop
    withTranslation: () => (Component) => {
        Component.defaultProps = { ...Component.defaultProps, t: (key) => key };
        return Component;
    },
}));

describe('Checkboxes', () => {

    const defaultChoiceValue = 'test1';
    const anotherChoiceValue = 'test2';
    const choiceValue3 = 'test3';
    const labelForChoice2 = 'label for test2';
    const labelForChoice3 = 'label for test3';
    const testChoices = [
        { value: defaultChoiceValue },
        { value: anotherChoiceValue, label: labelForChoice2 },
        { value: choiceValue3, label: labelForChoice3 },
    ];


    beforeEach(() => {
        mockOnChange.mockClear();
    });

    test('Default props', () => {
        const { container } = render(<InputCheckbox
            id = {testId}
            onValueChange = {mockOnChange}
            choices = {[]}
        /> as any);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('All props', () => {
        const { container } = render(<InputCheckbox
            id = {testId}
            onValueChange = {mockOnChange}
            value = {[anotherChoiceValue]}
            defaultValue = {[defaultChoiceValue]}
            choices = {testChoices}
            columns = {1}
            disabled = {false}
            localePrefix = "something"
            allowSelectAll = {false}
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('Disabled', () => {
        const { container } = render(<InputCheckbox
            id = {testId}
            onValueChange = {mockOnChange}
            value = {[anotherChoiceValue]}
            disabled = {true}
            choices = {testChoices}
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('With columns', () => {
        const { container } = render(<InputCheckbox
            id = {testId}
            onValueChange = {mockOnChange}
            choices = {testChoices}
            columns = {2}
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('allow select all', () => {
        const { container } = render(<InputCheckbox
            id = {testId}
            onValueChange = {mockOnChange}
            choices = {testChoices}
            columns = {2}
            allowSelectAll = {true}
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('Default value', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckbox
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {[defaultChoiceValue]}
                choices = {testChoices}
            />
        );
        const input = container.querySelector(`input#${testId}_${defaultChoiceValue}`) as HTMLInputElement;
        expect(input.checked).toBeTruthy();

        const input2 = container.querySelector(`input#${testId}_${anotherChoiceValue}`) as HTMLInputElement;
        expect(input2.checked).toBeFalsy();

        const input3 = container.querySelector(`input#${testId}_${choiceValue3}`) as HTMLInputElement;
        expect(input3.checked).toBeFalsy();
    });

    test('Default and initial value', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckbox
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {[defaultChoiceValue]}
                value = {[anotherChoiceValue]}
                choices = {testChoices}
            />
        );
        const input = container.querySelector(`input#${testId}_${defaultChoiceValue}`) as HTMLInputElement;
        expect(input.checked).toBeFalsy();

        const input2 = container.querySelector(`input#${testId}_${anotherChoiceValue}`) as HTMLInputElement;
        expect(input2.checked).toBeTruthy();

        const input3 = container.querySelector(`input#${testId}_${choiceValue3}`) as HTMLInputElement;
        expect(input3.checked).toBeFalsy();
    });

    test('Default and initial empty value', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckbox
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {[defaultChoiceValue]}
                value = {[]}
                choices = {testChoices}
            />
        );
        const input = container.querySelector(`input#${testId}_${defaultChoiceValue}`) as HTMLInputElement;
        expect(input.checked).toBeFalsy();

        const input2 = container.querySelector(`input#${testId}_${anotherChoiceValue}`) as HTMLInputElement;
        expect(input2.checked).toBeFalsy();

        const input3 = container.querySelector(`input#${testId}_${choiceValue3}`) as HTMLInputElement;
        expect(input3.checked).toBeFalsy();
    });

    test('Invalid default and value', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckbox
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {['not a value']}
                value = {['still not a choice']}
                choices = {testChoices}
            />
        );
        const input1 = container.querySelector(`input#${testId}_${defaultChoiceValue}`) as HTMLInputElement;
        expect(input1.checked).toBeFalsy();

        const input2 = container.querySelector(`input#${testId}_${anotherChoiceValue}`) as HTMLInputElement;
        expect(input2.checked).toBeFalsy();

        const input3 = container.querySelector(`input#${testId}_${choiceValue3}`) as HTMLInputElement;
        expect(input3.checked).toBeFalsy();
    });

    test('Call onChange', async () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckbox
                id = {testId}
                onValueChange = {mockOnChange}
                choices = {testChoices}
            />
        );
        // Make sure the select all button is not there
        expect(container.querySelector(`input#${testId}_selectAll`)).toBeNull();

        const defaultValueInput = container.querySelector(`input#${testId}_${defaultChoiceValue}`) as HTMLInputElement;
        expect(defaultValueInput.checked).toBeFalsy();
        // Manually set the DOM state to checked because the change simulation does not do it
        fireEvent.click(defaultValueInput);
        await TestUtils.flushPromises();
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        expect(mockOnChange).toHaveBeenCalledWith({ target: { value: [defaultChoiceValue] } });
    });

    test('Select/Unselect all', async () => {
        mockOnChange.mockClear();
        const { container, rerender } = render(
            <InputCheckbox
                id = {testId}
                onValueChange = {mockOnChange}
                choices = {testChoices}
                allowSelectAll = {true}
            />
        );
        const selectAllButton = container.querySelector(`input#${testId}_selectAll`) as HTMLInputElement;
        fireEvent.click(selectAllButton);
        await TestUtils.flushPromises();
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        expect(mockOnChange).toHaveBeenCalledWith({ target: { value: [defaultChoiceValue, anotherChoiceValue, choiceValue3] } });

        // Simulate parent changing the props to all selected, should unselect
        const newProps = {
            id: testId,
            onValueChange: mockOnChange,
            choices: testChoices,
            value: [defaultChoiceValue, anotherChoiceValue, choiceValue3],
            allowSelectAll: true
        };
        rerender(<InputCheckbox {...newProps} />);
        fireEvent.click(selectAllButton);
        await TestUtils.flushPromises();
        expect(mockOnChange).toHaveBeenCalledTimes(2);
        expect(mockOnChange).toHaveBeenCalledWith({ target: { value: [] } });
    });

});

describe('Boolean Checkboxes', () => {
    beforeEach(() => {
        mockOnChange.mockClear();
    });

    test('Default props', () => {
        const { container } = render(<InputCheckboxBoolean
            id = {testId}
            onValueChange = {mockOnChange}
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('All props', () => {
        const { container } = render(<InputCheckboxBoolean
            id = {testId}
            onValueChange = {mockOnChange}
            isChecked = {true}
            defaultChecked = {false}
            label = "testLabel"
            disabled = {false}
            localePrefix = "something"
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('Disabled', () => {
        const { container } = render(<InputCheckboxBoolean
            id = {testId}
            onValueChange = {mockOnChange}
            disabled = {true}
        />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('Default checked value', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckboxBoolean
                id = {testId}
                onValueChange = {mockOnChange}
                defaultChecked = {true}
                label = {testLabel}
            />
        );
        const input = container.querySelector(`input#${testId}_true`) as HTMLInputElement;
        expect(input.checked).toBeTruthy();
    });

    test('Default and initial checked values', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckboxBoolean
                id = {testId}
                onValueChange = {mockOnChange}
                defaultChecked = {true}
                isChecked = {false}
                label = {testLabel}
            />
        );
        const input = container.querySelector(`input#${testId}_true`) as HTMLInputElement;
        expect(input.checked).toBeFalsy();
    });

    test('Is checked value', () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckboxBoolean
                id = {testId}
                onValueChange = {mockOnChange}
                isChecked = {true}
                label = {testLabel}
            />
        );
        const input = container.querySelector(`input#${testId}_true`) as HTMLInputElement;
        expect(input.checked).toBeTruthy();
    });

    test('Call onChange', async () => {
        mockOnChange.mockClear();
        const { container } = render(
            <InputCheckboxBoolean
                id = {testId}
                onValueChange = {mockOnChange}
                label = {testLabel}
            />
        );
        const input = container.querySelector(`input#${testId}_true`) as HTMLInputElement;
        fireEvent.click(input);
        await TestUtils.flushPromises();
        expect(mockOnChange).toHaveBeenCalledTimes(1);
    });
});

