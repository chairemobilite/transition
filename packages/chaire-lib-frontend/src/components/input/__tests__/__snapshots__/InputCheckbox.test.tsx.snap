// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boolean Checkboxes All props 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column no-wrap"
  >
    <div
      class="tr__form-input-checkbox-container checked"
    >
      <div
        class="label-input-container"
      >
        <input
          checked=""
          class="_input-checkbox"
          id="SelectWidgetId_true"
          name="SelectWidgetId"
          type="checkbox"
          value="true"
        />
        <label
          for="SelectWidgetId_true"
        >
          <span>
            testLabel
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Boolean Checkboxes Default props 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column no-wrap"
  >
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          id="SelectWidgetId_true"
          name="SelectWidgetId"
          type="checkbox"
          value="true"
        />
        <label
          for="SelectWidgetId_true"
        >
          <span>
            Yes
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Boolean Checkboxes Disabled 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column no-wrap"
  >
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          disabled=""
          id="SelectWidgetId_true"
          name="SelectWidgetId"
          type="checkbox"
          value="true"
        />
        <label
          for="SelectWidgetId_true"
        >
          <span>
            Yes
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkboxes All props 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column no-wrap"
  >
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          id="SelectWidgetId_test1"
          name="SelectWidgetId"
          type="checkbox"
          value="test1"
        />
        <label
          for="SelectWidgetId_test1"
        >
          <span>
            something:test1
          </span>
        </label>
      </div>
    </div>
    <div
      class="tr__form-input-checkbox-container checked"
    >
      <div
        class="label-input-container"
      >
        <input
          checked=""
          class="_input-checkbox"
          id="SelectWidgetId_test2"
          name="SelectWidgetId"
          type="checkbox"
          value="test2"
        />
        <label
          for="SelectWidgetId_test2"
        >
          <span>
            label for test2
          </span>
        </label>
      </div>
    </div>
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          id="SelectWidgetId_test3"
          name="SelectWidgetId"
          type="checkbox"
          value="test3"
        />
        <label
          for="SelectWidgetId_test3"
        >
          <span>
            label for test3
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkboxes Default props 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column no-wrap"
  />
</div>
`;

exports[`Checkboxes Disabled 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column no-wrap"
  >
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          disabled=""
          id="SelectWidgetId_test1"
          name="SelectWidgetId"
          type="checkbox"
          value="test1"
        />
        <label
          for="SelectWidgetId_test1"
        >
          <span>
            main:test1
          </span>
        </label>
      </div>
    </div>
    <div
      class="tr__form-input-checkbox-container checked"
    >
      <div
        class="label-input-container"
      >
        <input
          checked=""
          class="_input-checkbox"
          disabled=""
          id="SelectWidgetId_test2"
          name="SelectWidgetId"
          type="checkbox"
          value="test2"
        />
        <label
          for="SelectWidgetId_test2"
        >
          <span>
            label for test2
          </span>
        </label>
      </div>
    </div>
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          disabled=""
          id="SelectWidgetId_test3"
          name="SelectWidgetId"
          type="checkbox"
          value="test3"
        />
        <label
          for="SelectWidgetId_test3"
        >
          <span>
            label for test3
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkboxes With columns 1`] = `
<div
  class="tr__form-input-checkbox-group-container"
>
  <div
    class="tr__form-input-checkbox-group-column"
  >
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          id="SelectWidgetId_test1"
          name="SelectWidgetId"
          type="checkbox"
          value="test1"
        />
        <label
          for="SelectWidgetId_test1"
        >
          <span>
            main:test1
          </span>
        </label>
      </div>
    </div>
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          id="SelectWidgetId_test2"
          name="SelectWidgetId"
          type="checkbox"
          value="test2"
        />
        <label
          for="SelectWidgetId_test2"
        >
          <span>
            label for test2
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    class="tr__form-input-checkbox-group-column"
  >
    <div
      class="tr__form-input-checkbox-container"
    >
      <div
        class="label-input-container"
      >
        <input
          class="_input-checkbox"
          id="SelectWidgetId_test3"
          name="SelectWidgetId"
          type="checkbox"
          value="test3"
        />
        <label
          for="SelectWidgetId_test3"
        >
          <span>
            label for test3
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkboxes allow select all 1`] = `
<div
  class="apptr__form-input-container"
>
  <div
    class="tr__form-input-checkbox-group-column"
  >
    <div
      class="label-input-container"
    >
      <input
        class="_input-checkbox"
        id="SelectWidgetId_selectAll"
        name="SelectWidgetId"
        type="button"
        value="main:SelectAll"
      />
    </div>
  </div>
  <div
    class="tr__form-input-checkbox-group-container"
  >
    <div
      class="tr__form-input-checkbox-group-column"
    >
      <div
        class="tr__form-input-checkbox-container"
      >
        <div
          class="label-input-container"
        >
          <input
            class="_input-checkbox"
            id="SelectWidgetId_test1"
            name="SelectWidgetId"
            type="checkbox"
            value="test1"
          />
          <label
            for="SelectWidgetId_test1"
          >
            <span>
              main:test1
            </span>
          </label>
        </div>
      </div>
      <div
        class="tr__form-input-checkbox-container"
      >
        <div
          class="label-input-container"
        >
          <input
            class="_input-checkbox"
            id="SelectWidgetId_test2"
            name="SelectWidgetId"
            type="checkbox"
            value="test2"
          />
          <label
            for="SelectWidgetId_test2"
          >
            <span>
              label for test2
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="tr__form-input-checkbox-group-column"
    >
      <div
        class="tr__form-input-checkbox-container"
      >
        <div
          class="label-input-container"
        >
          <input
            class="_input-checkbox"
            id="SelectWidgetId_test3"
            name="SelectWidgetId"
            type="checkbox"
            value="test3"
          />
          <label
            for="SelectWidgetId_test3"
          >
            <span>
              label for test3
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
